core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,165 @@
+        SERVER_TIMEOUT,
+        RESOLVER_TIMEOUT;
+
+        final String lowerCasedName;
+
+        DnsExceptionTypes() {
+            lowerCasedName = Ascii.toLowerCase(name());
+        }
+    }
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry,
+                                     DnsQuestion question, MeterIdPrefix prefix) {
+        this.meterRegistry = meterRegistry;
+        meterIdPrefix = prefix;
+        this.question = question;
+    }
+
+    @Override
+    public void queryWritten(InetSocketAddress dnsServerAddress, ChannelFuture future) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".written"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(SERVER_TAG, dnsServerAddress.getAddress().getHostAddress()))).increment();
+    }
+
+    @Override
+    public void queryCancelled(int queriesRemaining) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".cancelled"),
+                NAME_TAG, question.name()).increment();
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryRedirected(List<InetSocketAddress> nameServers) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".redirected"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()), Tag.of(SERVERS_TAG,
+                        nameServers.stream().map(addr -> addr.getAddress().getHostAddress())
+                                .collect(Collectors.joining(","))))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryCNAMEd(DnsQuestion cnameQuestion) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".cnamed"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(CNAME_TAG, cnameQuestion.name()))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryNoAnswer(DnsResponseCode code) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".noanswer"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(CODE_TAG, String.valueOf(code.intValue())))).increment();
+        return this;
+    }
+
+    @Override
+    public void queryFailed(Throwable cause) {
+        meterRegistry.counter(meterIdPrefix.name(),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()), Tag.of(RESULT_TAG, "failure"),
+                        Tag.of(CAUSE_TAG, determineDnsExceptionTag(cause).lowerCasedName))).increment();
+    }
+
+    @Override
+    public void querySucceed() {
+        meterRegistry.counter(meterIdPrefix.name(),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()), Tag.of(RESULT_TAG, "success"),
+                        Tag.of(CAUSE_TAG, "none"))).increment();
```

Comment:
Can we remove this `CAUSE_TAG`?

Reply:
Can we? `queryFailed()` uses the same meter name, so we have to keep the same set of tags.

Reply:
Oops, I missed that. 😅